teps/0154-concise-remote-resolver-syntax.md

```diff
@@ -0,0 +1,326 @@
+https://raw.githubusercontent.com/tektoncd-catalog/git-clone/main/task/git-clone/git-clone.yaml
+```
+
+Using the above example, a user can do the following when referencing the Task:
+```yaml
+taskRef:
+  resolver: http
+  name: "https://raw.githubusercontent.com/tektoncd-catalog/git-clone/main/task/git-clone/git-clone.yaml?http-username=git&http-password-secret=git-secret&http-password-secret-key=git-token"
+```
+
+##### Hub Resolver
+Based on `hub` resolver's [params](https://tekton.dev/docs/pipelines/hub-resolver/#parameters), we define the following `resolver-reference` format: 
+
+```
+artifacthub://<catalog>/<kind>/<name>@<version>
+tektonhub://<catalog>/<kind>/<name>@<version>
+```
+**Note** that we use a scheme called `artifacthub` or `tektonhub` which is only meaningful to the hub resolver.
+
+Example:
+
+```
+artifacthub://tekton-catalog-tasks/task/git-clone@0.6
+```
+
+Using the above example, a user can do the following when referencing the Task:
+
+```yaml
+taskRef:
+  resolver: hub
+  name: "artifacthub://tekton-catalog-tasks/task/git-clone@0.6"
+```
+
+##### Cluster Resolver
+Based on `cluster` resolver's [params](https://tekton.dev/docs/pipelines/cluster-resolver/#parameters), we define the following `resolver-reference` format: 
+
+```
+cluster://<namespace>/<kind>/<name>
+```
+
+**Note** that we use a scheme called `cluster` which is only meaningful to the cluster resolver.
+
+Example:
+
+```
+cluster://default/task/git-clone
+```
+
+Using the above example, a user can do the following when referencing the Task:
+
+```yaml
+taskRef:
+  resolver: "cluster"
+  name: "cluster://default/task/git-clone" 
+```
+
+##### Custom Resolver
+When implementing a new custom resolver, just like users have the freedom to choose their parameters, they also have the freedom to define their `name` (or `url`) based on:
+
+```
+<scheme>://<location>[@<version>][#<selector>][?<params>]
+```
+
+The custom resolver can choose the schemes it wants to subscribe to and for each scheme, it can define the remainder of `location, version, selectors and params` as needed. An example `scheme` could be `<custom-resolver-name>`. A `location` may be made up of concrete units e.g. `location = <foo>/<bar>` (i.e. expect two parts) that are up to the discretion of the custom resolvers.
+
+The custom resolver must validate the `name` which will be passed within the `Remote Resolution CRD's` spec. The `name` (or `url`) can then be parsed based on the format for that `scheme` and used to fetch the remote resource.
+
+### Stage 2
+
+In this stage, we enable the Tekton controller to be able to identify the resolvers on the clusters that have subscribed to the scheme of the incoming [resolver reference](#resolver-reference).
+
+#### Subscribing to `scheme`s
+
+Each remote resolver will subscribe to `scheme`s. e.g. the git resolver might subscribe to `scheme`s like `git+https`, `https` or a bundle resolver might subscribe to `scheme`s like `oci`.
```

Comment:
One question that I didn't though of before : what happens if the "other side" of the scheme is different from one provider to the other ? For example, for `https`, with the git resolver, the parameter is `secret`, but for the http resolver it is `http-password-secret` — how should we / the user handle that ?

Reply:
I think if users need to know what parameters they can pass, that is very specific to the resolvers since the names of the parameters are defined by the resolvers. 

In that scenario, users won't be able to try arbitrary parameters. Infact, they will have to know exactly which resolver they are calling and use parameters that the resolver understands.

On our end, we would try both resolvers subscribed to the scheme. The resolvers that do not understand the params from the scheme should error out. The resolvers should do this check when they write `ValidateUrl` that the resolver framework calls.

Reply:
Fair enough.